zip-0243.rst

```diff
@@ -0,0 +1,291 @@
+
+
+Abstract
+========
+
+This proposal defines a new transaction digest algorithm for signature verification from the Sapling network
+upgrade, to account for the presence of Sapling shielded inputs and outputs in transactions.
+
+
+Motivation
+==========
+
+The Sapling network upgrade introduced new shielded inputs and outputs. We want these to be covered by the
+transaction digest algorithm used for signatures, in order to ensure they are correctly bound.
+
+
+Specification
+=============
+
+A new transaction digest algorithm is defined::
+
+  BLAKE2b-256 hash of the serialization of:
+    1. header of the transaction (4-byte little endian)
+    2. nVersionGroupId of the transaction (4-byte little endian)
+    3. hashPrevouts (32-byte hash)
+    4. hashSequence (32-byte hash)
+    5. hashOutputs (32-byte hash)
+    6. hashJoinSplits (32-byte hash)
+    7. hashShieldedSpends (32-byte hash)
+    8. hashShieldedOutputs (32-byte hash)
+    9. nLockTime of the transaction (4-byte little endian)
+   10. nExpiryHeight of the transaction (4-byte little endian)
+   11. valueBalance of the transaction (8-byte little endian)
+   12. sighash type of the signature (4-byte little endian)
+
+   13. If we are serializing an input (i.e. this is not a JoinSplit signature hash):
+       a. outpoint (32-byte hash + 4-byte little endian) 
+       b. scriptCode of the input (serialized as scripts inside CTxOuts)
+       c. value of the output spent by this input (8-byte little endian)
+       d. nSequence of the input (4-byte little endian)
+
+The new algorithm is based on the transaction digest algorithm defined in ZIP 143 [#zip-0143]_.
+
+The new algorithm MUST be used for signatures created over the Sapling transaction format [#zip-0XXX]_.
+Combined with the new consensus rule that v3 transaction formats will be invalid from the Sapling upgrade
+[#zip-0XXX]_, this effectively means that all transaction signatures from the Sapling activation height will
+use the new algorithm [#zip-0XXX]_.
+
+The BLAKE2b-256 personalization field [#BLAKE2-personalization]_ is set to::
+
+  "ZcashSigHash" || CONSENSUS_BRANCH_ID
+
+``CONSENSUS_BRANCH_ID`` is the little-endian encoding of ``BRANCH_ID`` for the epoch of the block containing
+the transaction. [#zip-0200]_ Domain separation of the signature hash across parallel branches provides replay
+protection: transactions targeted for one branch will have invalid signatures on other branches.
+
+Transaction creators MUST specify the epoch they want their transaction to be mined in. Across a network
+upgrade, this means that if a transaction is not mined before the activation height, it will never be mined.
+
+Semantics of the original sighash types are as in ZIP 143 [#zip-0143]_.
+
+Field definitions
+-----------------
+
+The items 1, 2, 3, 4, 5, 9, 10, 12, and 13 have the same meaning as in ZIP 143 [#zip-0143]_.
+
+6: ``hashJoinSplits``
```

Comment:
An interesting question that came up in the discussion here  https://github.com/zcash/zcash/issues/2593#issuecomment-381322435
is whether when it's `SIGHASH_SINGLE` we might consider not including `hashJoinSplits,hashShieldedOutputs`.
This would preserve the intended meaning of that signing type in respect to the shielded outputs.
There are similar questions for the other sighash types.